tools/Packer/Program.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Text;
+
+namespace Packer
+{
+    public class Program
+    {
+        private const string ConfigFile = "packer.config.json";
+
+        public static void Main(string[] args)
+        {
+            var config = ConfigModel.FromFile(Path.Combine(Environment.CurrentDirectory, ConfigFile));
+
+            var tempDir = GetTempDirectory();
+
+            try
+            {
+                CloneRepo(config.CloneUrl, Path.Combine(tempDir, config.CloneDir));
+
+                foreach (var dependency in config.GitDependencies)
+                {
+                    CloneRepo(dependency.CloneUrl, Path.Combine(tempDir, dependency.CloneDir));
+                }
+
+                CreatePackage(tempDir, config);
+            }
+            catch (Exception e)
+            {
+                Console.Error.WriteLine(e.Message);
+            }
+            finally
+            {
+                RemoveReadOnlyAttribute(tempDir);
+                Directory.Delete(tempDir, true);
+            }
+        }
+
+        private static string GetTempDirectory()
+        {
+            var tempDirPath = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+            Directory.CreateDirectory(tempDirPath);
+            return tempDirPath;
+        }
+
+        private static void CreatePackage(string tempDir, ConfigModel config)
+        {
+            var packageName = Path.Combine(Environment.CurrentDirectory, $"{config.CloneDir}-{config.Version}.zip");
+
+            if (File.Exists(packageName))
+            {
+                File.Delete(packageName);
+            }
+
+            var archive = ZipFile.Open(packageName, ZipArchiveMode.Create);
+
+            var blackList = config.PathsToRemove
+                .Select(path => Path.Combine(tempDir, config.CloneDir, path))
+                .Concat(config.GitDependencies.SelectMany(dependency => dependency.PathsToRemove
+                    .Select(subPath => Path.Combine(tempDir, dependency.CloneDir, subPath))))
+                .ToHashSet();
+
+            AddToArchiveRecursive(archive, tempDir, blackList, tempDir.Length + 1);
+        }
+
+        private static void AddToArchiveRecursive(ZipArchive archive, string path, HashSet<string> blackList,
+            int basePathLength)
+        {
+            foreach (var file in Directory.GetFiles(path))
+            {
+                if (blackList.Contains(file))
+                {
+                    continue;
+                }
+
+                // Strip away the base path (temporary directory) length + the leading slash
+                var entry = archive.CreateEntry(file.Remove(0, basePathLength));
+                using (var stream = entry.Open())
+                {
+                    stream.Write(File.ReadAllBytes(file));
```

Comment:
One thing that may be missing is maintaining file attributes, for example "+x" on shell scripts.

Reply:
I don't think we have any shell scripts being zipped up - but will look into file attributes